src/lib/api/utils.ts

```diff
@@ -1,16 +1,124 @@
 import { GuildSettings } from '#lib/database';
 import { hasAtLeastOneKeyInMap } from '#utils/comparators';
-import { Guild, GuildMember, Permissions } from 'discord.js';
+import { Store } from '@sapphire/pieces';
+import type { ApiRequest, ApiResponse, LoginData } from '@sapphire/plugin-api';
+import { createFunctionInhibitor } from '@skyra/decorators';
+import type { RESTAPIPartialCurrentUserGuild } from 'discord-api-types/v8';
+import { Client, Guild, GuildMember, Permissions } from 'discord.js';
+import { RateLimitManager } from 'klasa';
+import { flattenGuild } from './ApiTransformers';
+import type { OauthFlattenedGuild, PartialOauthFlattenedGuild, TransformedLoginData } from './types';
 
 function isAdmin(member: GuildMember, roles: readonly string[]): boolean {
 	return roles.length === 0 ? member.permissions.has(Permissions.FLAGS.MANAGE_GUILD) : hasAtLeastOneKeyInMap(member.roles.cache, roles);
 }
 
-export async function canManage(guild: Guild, member: GuildMember) {
+export const authenticated = () =>
+	createFunctionInhibitor(
+		(request: ApiRequest) => Boolean(request.auth?.token),
+		(_request: ApiRequest, response: ApiResponse) => response.error(403)
+	);
+
+export function ratelimit(bucket: number, cooldown: number, auth = false) {
+	const manager = new RateLimitManager(bucket, cooldown);
+	const xRateLimitLimit = bucket;
+	return createFunctionInhibitor(
+		(request: ApiRequest, response: ApiResponse) => {
+			const id = (auth ? request.auth!.id : request.headers['x-forwarded-for'] || request.connection.remoteAddress) as string;
+			const bucket = manager.acquire(id);
+
+			response.setHeader('Date', new Date().toUTCString());
+			if (bucket.limited) {
+				response.setHeader('Retry-After', bucket.remainingTime.toString());
+				return false;
+			}
+
+			try {
+				bucket.drip();
+			} catch {}
+
+			response.setHeader('X-RateLimit-Limit', xRateLimitLimit);
+			response.setHeader('X-RateLimit-Remaining', bucket.bucket.toString());
+			response.setHeader('X-RateLimit-Reset', bucket.remainingTime.toString());
+
+			return true;
+		},
+		(_request: ApiRequest, response: ApiResponse) => {
+			response.error(429);
+		}
+	);
+}
+
+export async function canManage(guild: Guild, member: GuildMember): Promise<boolean> {
 	if (guild.ownerID === member.id) return true;
 	if (member.permissions.has(Permissions.FLAGS.MANAGE_GUILD)) return true;
 
 	const [roles, pnodes] = await guild.readSettings((settings) => [settings[GuildSettings.Roles.Admin], settings.permissionNodes]);
 
 	return isAdmin(member, roles) && (pnodes.run(member, 'conf') ?? false);
 }
+
+export async function getManageable(id: string, oauthGuild: RESTAPIPartialCurrentUserGuild, guild: Guild | undefined): Promise<boolean> {
+	if (oauthGuild.owner) return true;
+	if (typeof guild === 'undefined') return new Permissions(Number(oauthGuild.permissions)).has(Permissions.FLAGS.MANAGE_GUILD);
+
+	const member = await guild.members.fetch(id).catch(() => null);
```

Comment:
Note that we really do need to fetch. In dev I was missing various guilds because they weren't cached yet.